### ISSUE TITLE
install java 8.151 (and 9.0.1) fails

### ISSUE THREAD

**Issue by mrinalwadhwa** [skotchpine/asdf-java#7]
```
» asdf install java 8.151
---- ASDF JAVA MESSAGE ----
downloading: http://download.oracle.com/otn-pub/java/jdk/8u151-b12/e758a0de34e24606bca991d704f6dcbf/jdk-8u151-macosx-x64.dmg
---- ASDF JAVA ERROR ----
asdf java failed while downloading java dist
----
```

**Comment by mrinalwadhwa** [comment]
9.0.1 fails too, seems oracle.com doesn't allow the scripted download.
```
» asdf install java 9.0.1
---- ASDF JAVA MESSAGE ----
downloading: http://download.oracle.com/otn-pub/java/jdk/9.0.1+11/jdk-9.0.1_osx-x64_bin.dmg
---- ASDF JAVA ERROR ----
asdf java failed while downloading java dist
----
```

**Comment by skotchpine** [comment]
I've found this issue to be intermittent in the past. I'm doing to wait a few hours to a day before escalating.

A better solution here may be including openjdk versions and suggesting that as an alternative in the stderr message.

**Comment by brianvanburken** [comment]
I have the same issue. After digging into the issue I think I've found the problem. Version 8.151 is no longer downloadable. Though there is a newer version `8.161` now that could be added to this repository: http://download.oracle.com/otn-pub/java/jdk/8u161-b12/2f38c3b165be4555a1fa6e98c45e0808/jdk-8u161-macosx-x64.dmg.

Looking at the source code (https://github.com/skotchpine/asdf-java/blob/master/bin/install#L65) I see that the URL that is created for `9.x.x` is no longer valid. If you look at the latest version `9.0.4` (http://download.oracle.com/otn-pub/java/jdk/9.0.4+11/c2514751926b4512b076cc82f959763f/jdk-9.0.4_osx-x64_bin.dmg) they too have a hash in the URL as before which is missing for the `9.0.1` version.

Long story short: the current URL's are no longer available and must be replaced by new ones.

Hopefully this will help a bit :)

**Comment by croaky** [comment]
A pull request to add 8.161 only (does not remove / touch 8.151 or 9.0.1): https://github.com/skotchpine/asdf-java/pull/8

**Comment by skotchpine** [comment]
Thanks, all. I'm going to merge the two latest patches and review all this later this week. I'd like more intelligent reporting for these issues.

**Comment by skotchpine** [comment]
I've been procrastinating hard with this... XD

**Comment by brianvanburken** [comment]
No worries! It's just a lot of work to just keep track of all new versions and updating the references. You already did a awesome job building this plugin ❤️.

Maybe you could use a remote repository for versions? I couldn't find a official way to find all versions, but a solution could come from how other version managers solve this problem. For example, you could parse the Jabba list [https://github.com/shyiko/jabba/blob/master/index.json](https://github.com/shyiko/jabba/blob/master/index.json).

**Comment by skotchpine** [comment]
With things like this that we don't build locally, I think a repo is a good idea. But I'm not planning to dump capital into this just yet... :/ I'll fiddle with Oracle's site sometime soon.

**Comment by trustin** [comment]
It seems like it started to happen again? Can't install any of 8.161, 9.0.4 and 10. It worked just fine last night for me. :sob:

**Comment by trustin** [comment]
The following is the output of the `curl` command, ran with a `-a` option, which shows a 404 response:

```
$ asdf install java 8.161
---- ASDF JAVA MESSAGE ----
downloading: http://download.oracle.com/otn-pub/java/jdk/8u161-b12/2f38c3b165be4555a1fa6e98c45e0808/jdk-8u161-linux-x64.tar.gz
*   Trying 118.214.65.232...
* TCP_NODELAY set
* Connected to download.oracle.com (118.214.65.232) port 80 (#0)
> GET /otn-pub/java/jdk/8u161-b12/2f38c3b165be4555a1fa6e98c45e0808/jdk-8u161-linux-x64.tar.gz HTTP/1.1
> Host: download.oracle.com
> User-Agent: curl/7.59.0
> Accept: */*
> Cookie: oraclelicense=a
> 
< HTTP/1.1 302 Moved Temporarily
< Server: AkamaiGHost
< Content-Length: 0
< Location: https://edelivery.oracle.com/otn-pub/java/jdk/8u161-b12/2f38c3b165be4555a1fa6e98c45e0808/jdk-8u161-linux-x64.tar.gz
< Date: Wed, 18 Apr 2018 10:55:12 GMT
< Connection: keep-alive
< 
* Connection #0 to host download.oracle.com left intact
* Issue another request to this URL: 'https://edelivery.oracle.com/otn-pub/java/jdk/8u161-b12/2f38c3b165be4555a1fa6e98c45e0808/jdk-8u161-linux-x64.tar.gz'
*   Trying 104.74.178.67...
* TCP_NODELAY set
* Connected to edelivery.oracle.com (104.74.178.67) port 443 (#1)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
  CAfile: /etc/ssl/certs/ca-certificates.crt
  CApath: none
} [5 bytes data]
* TLSv1.2 (OUT), TLS handshake, Client hello (1):
} [218 bytes data]
* TLSv1.2 (IN), TLS handshake, Server hello (2):
{ [108 bytes data]
* TLSv1.2 (IN), TLS handshake, Certificate (11):
{ [3889 bytes data]
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
{ [333 bytes data]
* TLSv1.2 (IN), TLS handshake, Server finished (14):
{ [4 bytes data]
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
} [70 bytes data]
* TLSv1.2 (OUT), TLS change cipher, Client hello (1):
} [1 bytes data]
* TLSv1.2 (OUT), TLS handshake, Finished (20):
} [16 bytes data]
* TLSv1.2 (IN), TLS handshake, Finished (20):
{ [16 bytes data]
* SSL connection using TLSv1.2 / ECDHE-RSA-AES256-GCM-SHA384
* ALPN, server accepted to use http/1.1
* Server certificate:
*  subject: C=US; ST=California; L=Redwood Shores; O=Oracle Corporation; OU=Content Management Services IT; CN=www.oracle.com
*  start date: Mar 25 00:00:00 2018 GMT
*  expire date: Apr 23 12:00:00 2019 GMT
*  subjectAltName: host "edelivery.oracle.com" matched cert's "edelivery.oracle.com"
*  issuer: C=US; O=DigiCert Inc; OU=www.digicert.com; CN=GeoTrust RSA CA 2018
*  SSL certificate verify ok.
} [5 bytes data]
> GET /otn-pub/java/jdk/8u161-b12/2f38c3b165be4555a1fa6e98c45e0808/jdk-8u161-linux-x64.tar.gz HTTP/1.1
> Host: edelivery.oracle.com
> User-Agent: curl/7.59.0
> Accept: */*
> Cookie: oraclelicense=a
> 
{ [5 bytes data]
< HTTP/1.1 302 Moved Temporarily
< Server: Oracle-HTTP-Server-11g
< Location: http://download.oracle.com/otn-pub/java/jdk/8u161-b12/2f38c3b165be4555a1fa6e98c45e0808/jdk-8u161-linux-x64.tar.gz?AuthParam=1524049032_76ff65dcb8d0f3501ace68ac0077ad3e
< Content-Length: 531
< Content-Type: text/html
< Content-Language: en
< Date: Wed, 18 Apr 2018 10:55:12 GMT
< Connection: keep-alive
< 
* Ignoring the response-body
{ [531 bytes data]
* Connection #1 to host edelivery.oracle.com left intact
* Issue another request to this URL: 'http://download.oracle.com/otn-pub/java/jdk/8u161-b12/2f38c3b165be4555a1fa6e98c45e0808/jdk-8u161-linux-x64.tar.gz?AuthParam=1524049032_76ff65dcb8d0f3501ace68ac0077ad3e'
* Found bundle for host download.oracle.com: 0x5613b5ca9030 [can pipeline]
* Re-using existing connection! (#0) with host download.oracle.com
* Connected to download.oracle.com (118.214.65.232) port 80 (#0)
> GET /otn-pub/java/jdk/8u161-b12/2f38c3b165be4555a1fa6e98c45e0808/jdk-8u161-linux-x64.tar.gz?AuthParam=1524049032_76ff65dcb8d0f3501ace68ac0077ad3e HTTP/1.1
> Host: download.oracle.com
> User-Agent: curl/7.59.0
> Accept: */*
> Cookie: oraclelicense=a
> 
* The requested URL returned error: 404 Not Found
* stopped the pause stream!
* Closing connection 0
```

**Comment by whaley** [comment]
URLs have changed with new versions.  Updating bin/install in my local copy to have 10.0.1 point at download.oracle.com/otn-pub/java/jdk/10.0.1+10/fb4372174a714e6b8c52526dc134031e/jdk-10.0.1_osx-x64_bin.dmg now works.

**Comment by skotchpine** [comment]
Closing this.
